Add documentation for object metadata and data caching (#587)
awslabs/mountpoint-s3: doc/CONFIGURATION.md

We've tried hard to make this simple command adopt good defaults for most scenarios. However, some scenarios may need additional configuration. This document shows how to configure these elements of Mountpoint:
* [AWS credentials](#aws-credentials)

> * [Caching configuration](#caching-configuration)

Reviewer: Should be below bucket configuration, and probably below file system configuration too. Also needs a sentence description (caching what?).

Mountpoint also respects access control lists (ACLs) applied to objects in your S3 bucket, but does not allow you to automatically attach ACLs to objects created with Mountpoint. A majority of modern use cases in Amazon S3 no longer require the use of ACLs. We recommend that you keep ACLs disabled for your S3 bucket, and instead use bucket policies to control access to your objects.

## Caching configuration

> Mountpoint now offers caching of metadata and object content allowing for reduced requests when reading files.

Reviewer: Documentation is independent of time, so no "now".

```suggestion
Mountpoint can optionally cache object metadata and content to reduce cost and improve performance for repeated reads to the same file.
```

Author: makes sense.

I was planning to put "Since v1.2.0, Mountpoint can optionally cache...", but replaced it with "now" as a placeholder. Exclude that?

This is particularly useful when reading the same files many times for the same Mountpoint filesystem.

To enable caching, use the `--caching` command-line flag.
This alone will enable caching of metadata using a default time-to-live (TTL) of 60 minutes.
Caching of object/file content on disk can also be enabled

> by providing a caching location using the `--data-cache-directory <DIR>` command-line flag.

Reviewer: Naming bikeshedding: `--cache-metadata`/`--cache-data <DIR>`?

And then mention that `--cache-data` implies `--cache-metadata`.

Reviewer 2: If we are starting with a single flag, would it make sense to have something like `--cache <DIR>`, which enables both metadata and data cache using `<DIR>` for the latter?

We can always add more specific flags like `--cache-metadata` and `--cache-data <DIR>` if we want to give users more detailed options, without breaking changes or ambiguities.

Author: I updated to use `--cache <DIR>`. We should in a separate PR make an update to the final CLI args which also adopts this (or makes whatever change we decide to make).

> Enabling caching relaxes the strong read-after-write consistency offered by Mountpoint with default configuration.
> Please read more in the [consistency and concurrency section of the semantics documentaton](./SEMANTICS.md#consistency-and-concurrency).

Reviewer: ```suggestion
Caching relaxes the strong read-after-write consistency offered by Amazon S3 and Mountpoint in its default configuration. See the [consistency and concurrency section of the semantics documentaton](./SEMANTICS.md#consistency-and-concurrency) for more details.
```

Reviewer: Also, maybe this should use the fancy `> [!WARNING]` thing.

Mountpoint caching can be further configured,
such as adjusting the metadata time-to-live (TTL) or the maximum space allowed to be used by the data cache.

> Review the caching options available using `mount-s3 --help`.

Reviewer: Move this up to the paragraph about the command-line flags.

WantedBy=remote-fs.target
```

## Caching configuration

Mountpoint can optionally cache object metadata and content to reduce cost and improve performance for repeated reads to the same file.

To enable caching, use the `--cache <CACHE_DIR>` command-line flag.
This flag will enable caching of metadata using a default time-to-live (TTL) of 60 minutes.
Object/file content will also be cached within the cache directory specified.

> Mountpoint will create a directory within this path, removing any files or directories within it at mount time and exit.

Reviewer: Don't think we should be deleting at exit time — that's something we can't change our mind about if we want to default to reusable cache later.

Author: I'm not sure what to pick here. Two paths forward.

- We don't do cleanup at mount or exit.
  - There is a risk that we fill up the disk with stale cache content that we don't track, since we don't attempt to ingest old cache data. We eventually get to a point where we have very little space remaining for the real cache data, unless the customer takes manual action.
  - We also provide a `--delete-cache` option which will delete the cache at mount and exit.
  - Note, we haven't defined our recommendations to the user. We haven't tested disk reuse, and I assume we should discourage it here in the documentation until we are happy that cache reuse is both well tested and supported.
- We do cleanup at mount and exit.
  - We mitigate the risk from the disk filling up.
  - We don't need to guide the customer on cache reuse, since its not possible to use it.
  - Allowing cache reuse later would potentially be a breaking change since we'd change the disk usage characteristics, and would likely need to be opt-in through some `--allow-cache-reuse`/`--no-delete-cache` flag.

I do think I actually prefer the former option, but we need to guide users in the right direction to ensure they aren't reusing the cache.

Author: My other thinking is around how we should cleanup the cache.

At the moment, we are planning to delete the `mp-cache-V1` directory. I think this is fine, but in this case I think we also need to guide users to using a unique value for `--cache <CACHE_DIR>` for different buckets so that the caches don't trample over each other. The other option is that we include a hash of the bucket name into the path (which I think we should do anyway) and _delete only that sub-directory, leaving any other buckets intact_. I don't see much issue with this either actually.

Author: Discussed on a call with @jamesbornholt and @passaro, documenting here:

- We'll implement the cache directory argument as a directory we are creating a sub-directory within, and deleting the contents at mount and at exit. This avoids issues where requiring a non-empty directory causes availability issues on unexpected process restarts, or with not deleting and disk space accumulating and not being cleaned.
- Configuration documentation will recommend to use a unique value for this argument when running multiple processes, to avoid any clobbering of each process' cache. If Mountpoint is used with the same path, it may cause performance issues when the cache is clobbered but we also expect mounting and un-mounting on a process to be relatively infrequent and the impact here should be limited.
- If we were to consider cache re-use in the future, its unlikely we can silently change behavior to allow re-use and we'd probably need a new argument for that anyway.

WantedBy=remote-fs.target
```

## Caching configuration

Mountpoint can optionally cache object metadata and content to reduce cost and improve performance for repeated reads to the same file.

> To enable caching, use the `--cache <CACHE_DIR>` command-line flag.

Reviewer: ```suggestion
To enable caching, use the `--cache <CACHE_DIR>` command-line flag, specifying the directory in which to store cached object content.
```

> This flag will enable caching of metadata using a default time-to-live (TTL) of 1 second.
> Object content will also be cached within the cache directory specified.

Reviewer: ```suggestion
This flag will also enable caching of metadata using a default time-to-live (TTL) of 1 second.
```

> Mountpoint will create a directory within this path, removing any files or directories within it at mount time and exit.

Reviewer: ```suggestion
Mountpoint will create a new subdirectory within the path that you specify, and will remove any existing files or directories within that subdirectory at mount time and at exit.
```

> When available space for the cache on disk is limited,
> content of files least recently accessed will be evicted to allow new data to be cached.

Reviewer: ```suggestion
By default, Mountpoint will limit the maximum size of the cache such that the free space on the file system does not fall below 5%, and will automatically evict the least recently used content from the cache when caching new content. You can instead manually configure the maximum size of the cache with the `--max-cache-size <MiB>` command-line argument.
```

When running multiple Mountpoint processes on the same host,
you should use unique cache directories to avoid different processes interfering with the others' cache content.

Mountpoint caching can be further configured,
such as adjusting the metadata time-to-live (TTL) or the maximum space allowed to be used by the data cache.

> Review the caching options available using `mount-s3 --help`.

Reviewer: Don't point to `--help` in documentation, this _is_ the documentation :-) Just describe the options you can configure.

WantedBy=remote-fs.target
```

## Caching configuration

Mountpoint can optionally cache object metadata and content to reduce cost and improve performance for repeated reads to the same file.

To enable caching, use the `--cache <CACHE_DIR>` command-line flag.
This flag will enable caching of metadata using a default time-to-live (TTL) of 1 second.
Object content will also be cached within the cache directory specified.
Mountpoint will create a directory within this path, removing any files or directories within it at mount time and exit.
When running multiple Mountpoint processes on the same host,
you should use unique cache directories to avoid different processes interfering with the others' cache content.

Mountpoint caching can be further configured,
such as adjusting the metadata time-to-live (TTL) or the maximum space allowed to be used by the data cache.

> Review the caching options available using `mount-s3 --help`.

Author: It may be worth calling out the metadata TTL specifically here also, since we expect this to be something customers may want to tune.

> [!WARNING]
> If you enable caching, Mountpoint will persist unencrypted object content from your S3 bucket at the location provided at mount.
> In order to protect your data, we recommend you restrict access to the data cache location.

### Using multiple Mountpoint processes on a host

The cache directory is not reusable by other Mountpoint processes and will be cleaned at mount time and exit.
When running multiple Mountpoint processes concurrently on the same host,
you should use unique cache directories to avoid different processes interfering with the others' cache content.

### Caching object content to random access memory

Mountpoint can instead cache object content to random access memory (RAM) using a RAM disk.

To create a RAM disk on Linux, you can use [tmpfs](https://www.kernel.org/doc/html/latest/filesystems/tmpfs.html) as shown below.
Replace the mount directory for tmpfs if required.

```
sudo mkdir /mnt/mp-cache-tmpfs
sudo mount -o uid=$(id --user),mode=700 -t tmpfs none /mnt/mp-cache-tmpfs
```

This will create a RAM disk mounted at `/mnt/mp-cache-tmpfs` with access restricted to the current user.

> By default, Linux limits it's size to 50% of the physical memory available on the system.

Reviewer 2: ```suggestion
By default, Linux limits its size to 50% of the physical memory available on the system.
```

The size is configurable using the `size` option.

You can then start Mountpoint using the directory where the RAM disk was mounted.

```

> mount-s3 DOC-EXAMPLE-BUCKET /path/to/mount --cache /mnt/tmpfs

Reviewer 2: ```suggestion
mount-s3 DOC-EXAMPLE-BUCKET /path/to/mount --cache /mnt/mp-cache-tmpfs
```
